src/messaging/ExchangeContext.h

```diff
@@ -55,7 +53,8 @@ class ExchangeContextDeletor
  *    It defines methods for encoding and communicating CHIP messages within an ExchangeContext
  *    over various transport mechanisms, for example, TCP, UDP, or CHIP Reliable Messaging.
  */
-class DLL_EXPORT ExchangeContext : public ReferenceCounted<ExchangeContext, ExchangeContextDeletor, 0>
+class DLL_EXPORT ExchangeContext : public ReliableMessageContext,
+                                   public ReferenceCounted<ExchangeContext, ExchangeContextDeletor, 0>
```

Comment:
I don't think derive from `ReliableMessageContext` is a good idea. I understand that it reduce a pointer from `ReliableMessageContext` to `ExchangeContext`. But I think it is better using `offsetof` and `container_of` macro instead of derive from it. Then `ReliableMessageContext::GetExchangeContext` can be implemented by: 

```
ExchangeContext * ReliableMessageContext::GetExchangeContext()
{
    return static_cast<ExchangeContext*>(this - offsetof(ExchangeContext, mReliableMessageContext));
}
```

I want to hear from @andy31415  and @bzbarsky-apple, that are we allowed to use macro like `offsetof` and `container_of`

Reply:
Do 'offsetof' or 'container_of' need we always declare  mReliableMessageContext at the beginning of ExchangeContext? if so, it is hard to maintain and easy to break if somebody else modify class ExchangeContext

Reply:
Using offsetof is just fine from my point of view, though that specific line is wrong (need to cast to a type that is one-byte-sized to subtract a number of bytes like offsetof).  If we're careful to `static_assert` things about `is_standard_layout` behavior is perfectly well-defined.

Reply:
why this method is better than inheritance? Inheritance also allow ExchangeContext to reuse mFlags field of ReliableMessageContext